src/ucp/core/ucp_context.h

```diff
@@ -441,45 +441,64 @@ static UCS_F_ALWAYS_INLINE int ucp_memory_type_cache_is_empty(ucp_context_h cont
             !context->memtype_cache->pgtable.num_regions);
 }
 
-static UCS_F_ALWAYS_INLINE ucs_memory_type_t
-ucp_memory_type_detect(ucp_context_h context, const void *address, size_t length)
+static UCS_F_ALWAYS_INLINE void
+ucp_memory_info_set_host(ucs_memory_info_t *mem_info)
+{
+    mem_info->type    = UCS_MEMORY_TYPE_HOST;
+    mem_info->sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_memory_detect(ucp_context_h context, const void *address, size_t length,
+                  ucs_memory_info_t *mem_info)
 {
-    ucs_memory_type_t mem_type;
     ucs_status_t status;
 
     if (ucs_likely(context->num_mem_type_detect_mds == 0)) {
-        return UCS_MEMORY_TYPE_HOST;
+        goto out_host_mem;
     }
 
     if (ucs_likely(context->memtype_cache != NULL)) {
         if (!context->memtype_cache->pgtable.num_regions) {
-            return UCS_MEMORY_TYPE_HOST;
+            goto out_host_mem;
         }
 
         status = ucs_memtype_cache_lookup(context->memtype_cache, address,
-                                          length, &mem_type);
-        if (status != UCS_OK) {
+                                          length, mem_info);
+        if (ucs_likely(status != UCS_OK)) {
             ucs_assert(status == UCS_ERR_NO_ELEM);
-            return UCS_MEMORY_TYPE_HOST;
+            goto out_host_mem;
         }
 
-        if (mem_type != UCS_MEMORY_TYPE_LAST) {
-            return mem_type;
+        if ((mem_info->type != UCS_MEMORY_TYPE_LAST) &&
+            ((mem_info->sys_dev != UCS_SYS_DEVICE_ID_UNKNOWN) ||
+             (mem_info->type == UCS_MEMORY_TYPE_HOST))) {
+            return;
         }
 
-        /* mem_type is UCS_MEMORY_TYPE_LAST: fall thru to memory detection by
-         * UCT memory domains */
+        /* Fall thru to slow-path memory detection by UCT memory domains */
     }
 
-    return ucp_memory_type_detect_mds(context, address, length);
+    ucp_memory_detect_slowpath(context, address, length, mem_info);
+    return;
+
+out_host_mem:
+    ucp_memory_info_set_host(mem_info);
 }
 
 static UCS_F_ALWAYS_INLINE ucs_memory_type_t
 ucp_get_memory_type(ucp_context_h context, const void *address,
                     size_t length, ucs_memory_type_t memory_type)
 {
-    return (memory_type == UCS_MEMORY_TYPE_UNKNOWN) ?
-           ucp_memory_type_detect(context, address, length) : memory_type;
+    ucs_memory_info_t mem_info;
+
+    /* TODO extract memory type from param */
+    if (memory_type == UCS_MEMORY_TYPE_UNKNOWN) {
```

Comment:
Shouldn't we throw an error if memory_type > TYPE_UNKNOWN? Or do we do parameter error checking in the layers above? 

Also, if user passes incorrect but valid type, it would be useful to have a configure option to do memory checking and warn if the user is passing the wrong type for the given buffer as part of nbx ops.

Reply:
added parameter checks